api/rosetta/data_block_integration_test.go

```diff
@@ -0,0 +1,685 @@
+
+			rec := httptest.NewRecorder()
+			ctx := echo.New().NewContext(req, rec)
+
+			err := api.Block(ctx)
+			assert.Error(t, err)
+
+			echoErr, ok := err.(*echo.HTTPError)
+			require.True(t, ok)
+
+			assert.Equal(t, http.StatusBadRequest, echoErr.Code)
+
+			gotErr, ok := echoErr.Message.(rosetta.Error)
+			require.True(t, ok)
+
+			assert.Equal(t, configuration.ErrorInvalidFormat, gotErr.ErrorDefinition)
+			assert.NotEmpty(t, gotErr.Description)
+		})
+	}
+
+}
+
+// blockRequest generates a BlockRequest with the specified parameters.
+func blockRequest(header flow.Header) rosetta.BlockRequest {
+
+	return rosetta.BlockRequest{
+		NetworkID: defaultNetworkID(),
+		BlockID: identifier.Block{
+			Index: header.Height,
+			Hash:  header.ID().String(),
+		},
+	}
+}
+
+func validateTransfer(t *testing.T, hash string, from string, to string, amount int64) transactionValidationFn {
+
+	t.Helper()
+
+	return func(tx *object.Transaction) {
+
+		assert.Equal(t, tx.ID.Hash, hash)
+		assert.Equal(t, len(tx.Operations), 2)
+
+		relatedOperations := make(map[uint]uint)
+
+		for _, op := range tx.Operations {
+
+			// save related operation IDs in a map so we can cross reference them
+			// TODO: check - is this special-casing the test too much perhaps?
+			// operation can have multiple related operations,
+			// but also this function does state that it's verifying a single transfer
+
+			if assert.Len(t, op.RelatedIDs, 1) {
+				relatedOperations[op.ID.Index] = op.RelatedIDs[0].Index
+			}
+
+			// validate operation and status
+			assert.Equal(t, op.Type, dps.OperationTransfer)
+			assert.Equal(t, op.Status, dps.StatusCompleted)
+
+			// validate currency
+			assert.Equal(t, op.Amount.Currency.Symbol, dps.FlowSymbol)
+			assert.Equal(t, op.Amount.Currency.Decimals, uint(dps.FlowDecimals))
+
+			// validate address
+			address := op.AccountID.Address
+			if address != from && address != to {
+				t.Errorf("unexpected account address (%v)", address)
+			}
+
+			// validate transfered amount
+			wantValue := strconv.FormatInt(amount, 10)
+			if address == from {
+				wantValue = "-" + wantValue
+			}
+
+			assert.Equal(t, op.Amount.Value, wantValue)
+		}
+
+		// cross-reference related operations - verify that the related operation backlinks to the original one
+		for id, relatedID := range relatedOperations {
+			assert.Contains(t, relatedOperations, relatedID)
+			assert.Equal(t, id, relatedOperations[relatedID])
+		}
+	}
+}
+
+func validateBlockID(t *testing.T, height uint64, hash string) blockIDValidationFn {
+
+	t.Helper()
+
+	return func(blockID identifier.Block) {
+		assert.Equal(t, height, blockID.Index)
+		assert.Equal(t, hash, blockID.Hash)
+	}
+}
+
+func validatorFromHeader(t *testing.T, header flow.Header) blockIDValidationFn {
+	return validateBlockID(t, header.Height, header.ID().String())
+}
+
+func rosettaTime(t time.Time) int64 {
+	return t.UnixNano() / 1_000_000
+}
+
+func knownHeaders(height uint64) flow.Header {
```

Comment:
I ended up returning headers made "by hand". Feels more correct to have them explicitly defined instead of relying on another piece of code (e.g. indexer) to retrieve them, even if it is ugly to see.

I would personally hide it in a third test file, though it would change the current `data_<entpoint>_integration_test.go` filename pattern.

Also hate how the `flow.HashToID([]byte{0x01, 0x02, 0x03, ...})` looks. However the `flow.HexStringToIdentifier()` returns two values, so couldn't init the struct nicely with that.

Bleh.

Reply:
The explicit declaration makes perfect sense. Ideally, we will be able to re-use these. That is also what I meant with "snapshot documentation". We should know what's in the snapshot, such as these headers.

Seeing how the `flow.Identifier` type is just a byte array, why don't you just do this?

```
ParentID: flow.Identifier{0xd4, 0x7b, ...}
```